File: src/data/rules/statBenefits.ts
/**
 * Armor 3.0 stat benefit curves (0-200 scale), transcribed from Bungie's
 * Armor 3.0 documentation and the 9.7.0 patch notes. Every stat scales
 * linearly per point; 101-200 grants an enhanced benefit that also scales
 * linearly, reaching the listed maximum at 200. Sources cached in
 * `src/data/meta/sources/`.
 */

export type ArmorStatName =
  | "Health"
  | "Melee"
  | "Grenade"
  | "Super"
  | "Class"
  | "Weapons";

export const ARMOR_STAT_NAMES: readonly ArmorStatName[] = [
  "Health",
  "Melee",
  "Grenade",
  "Super",
  "Class",
  "Weapons",
];

export const STAT_MAX = 200;
export const ENHANCED_THRESHOLD = 100;

export interface ScalingBenefit {
  /** Template with `{v}` placeholder, e.g. "+{v}% grenade ability damage". */
  template: string;
  /** Value reached at stat 100 (base range) or 200 (enhanced range). */
  max: number;
  /** Decimal places when rendering the interpolated value. */
  precision?: number;
}

export interface StatBenefitDefinition {
  stat: ArmorStatName;
  /** Always-on effects across 0-100, shown as static text. */
  baseEffects: readonly string[];
  /** Benefits that scale across 0-100 (value at stat 100 = max). */
  baseScaling: readonly ScalingBenefit[];
  /** Benefits that scale across 101-200 (value at stat 200 = max). */
  enhancedScaling: readonly ScalingBenefit[];
  /** Non-scaling notes for the enhanced range. */
  enhancedNotes: readonly string[];
}

export const STAT_BENEFITS: Record<ArmorStatName, StatBenefitDefinition> = {
  Health: {
    stat: "Health",
    baseEffects: [],
    baseScaling: [
      { template: "+{v} HP healing when picking up an Orb of Power", max: 70 },
      { template: "+{v}% flinch resistance", max: 10 },
    ],
    enhancedScaling: [
      { template: "+{v} shield capacity vs combatants", max: 20 },
      { template: "shields start recharging {v}% sooner", max: 25 },
      { template: "{v}% less time to fully recharge shields", max: 50 },
    ],
    enhancedNotes: [],
  },
  Melee: {
    stat: "Melee",
    baseEffects: [
      "Improves melee ability base cooldown",
      "Improves melee energy gained from external sources and regen scalars",
    ],
    baseScaling: [],
    enhancedScaling: [
      { template: "+{v}% melee ability damage", max: 30 },
    ],
    enhancedNotes: ["Applies to powered, unpowered, and glaive melees"],
  },
  Grenade: {
    stat: "Grenade",
    baseEffects: [
      "Improves grenade ability base cooldown",
      "Improves grenade energy gained from external sources and regen scalars",
    ],
    baseScaling: [],
    enhancedScaling: [
      { template: "+{v}% grenade ability damage", max: 65 },
    ],
    enhancedNotes: [],
  },
  Super: {
    stat: "Super",
    baseEffects: [
      "Improves Super energy gained from damaging targets (base cooldown unchanged)",
      "Improves Super energy gained from external sources and regen scalars",
    ],
    baseScaling: [],
    enhancedScaling: [
      { template: "+{v}% Super ability damage", max: 45 },
    ],
    enhancedNotes: [],
  },
  Class: {
    stat: "Class",
    baseEffects: [
      "Improves class ability base cooldown",
      "Improves class energy gained from external sources and regen scalars",
    ],
    baseScaling: [],
    enhancedScaling: [
      { template: "{v} HP overshield on casting your class ability (PvE)", max: 40 },
      { template: "{v} HP overshield on cast in PvP", max: 10 },
    ],
    enhancedNotes: [
      "Overshield duration is tied to the ability's cooldown length",
    ],
  },
  Weapons: {
    stat: "Weapons",
    baseEffects: ["Improves weapon reload and handling speeds"],
    baseScaling: [
      { template: "+{v}% damage vs minor and major combatants (primary and special)", max: 15 },
      { template: "+{v}% damage vs minor and major combatants (heavy)", max: 10 },
    ],
    enhancedScaling: [
      { template: "+{v}% damage vs bosses (primary and special, PvE)", max: 15 },
      { template: "+{v}% damage vs bosses (heavy, PvE)", max: 10 },
      { template: "+{v}% weapon damage vs Guardians (PvP)", max: 6 },
    ],
    enhancedNotes: [
      "Ammo bricks have a chance to contain more ammo than normal",
    ],
  },
};

/**
 * 9.7.0 sandbox notes layered on top of the curves (patch notes "Abilities"):
 * several damaging class abilities additionally scale with Class above 100,
 * and ability cooldown scalars were trimmed.
 */
export const CLASS_STAT_SCALING_ABILITIES: readonly string[] = [
  "Shieldburst",
  "Ascension",
  "Threaded Specter",
  "Drengr's Lash",
];

export const ABILITY_ECONOMY_NOTES: readonly string[] = [
  "Max passive ability-cooldown bonus from stats reduced by ~20% (9.7.0)",
  "Max active ability-cooldown bonus reduced from 190% to 125% (9.7.0)",
  "Super energy from boss damage reduced by 60% (9.7.0)",
  "Grenade/Melee stat boss-damage bonuses are additive, not multiplicative (9.7.0)",
];

function interpolate(value: number, rangeStart: number, max: number): number {
  const clamped = Math.min(Math.max(value - rangeStart, 0), 100);
  return (clamped / 100) * max;
}

function renderBenefit(benefit: ScalingBenefit, amount: number): string {
  const rendered = amount.toFixed(benefit.precision ?? 0);
  return benefit.template.replace("{v}", rendered);
}

/**
 * Computes the human-readable benefit lines for a stat at a target value,
 * e.g. computeBenefitsAt("Melee", 200) includes "+30% melee ability damage".
 * Enhanced benefits are omitted at or below 100.
 */
export function computeBenefitsAt(
  stat: ArmorStatName,
  value: number,
): string[] {
  const definition = STAT_BENEFITS[stat];
  const lines: string[] = [...definition.baseEffects];
  for (const benefit of definition.baseScaling) {
    lines.push(renderBenefit(benefit, interpolate(value, 0, benefit.max)));
  }
  if (value <= ENHANCED_THRESHOLD) return lines;
  for (const benefit of definition.enhancedScaling) {
    lines.push(renderBenefit(benefit, interpolate(value, 100, benefit.max)));
  }
  lines.push(...definition.enhancedNotes);
  return lines;
}

File: packages/sandbox_data/lib/src/stat_benefits.dart
import 'armor_stat_name.dart';

/// Armor 3.0 stat benefit curves (0–200 scale), transcribed from Bungie's
/// Armor 3.0 documentation and the 9.7.0 patch notes. Port of
/// `src/data/rules/statBenefits.ts`.

const int statMax = 200;
const int enhancedThreshold = 100;

/// Template with `{v}` placeholder and max value at range end.
class ScalingBenefit {
  const ScalingBenefit({
    required this.template,
    required this.max,
    this.precision = 0,
  });

  /// e.g. "+{v}% grenade ability damage".
  final String template;

  /// Value reached at stat 100 (base) or 200 (enhanced).
  final num max;

  /// Decimal places when rendering the interpolated value.
  final int precision;
}

class StatBenefitDefinition {
  const StatBenefitDefinition({
    required this.stat,
    this.baseEffects = const [],
    this.baseScaling = const [],
    this.enhancedScaling = const [],
    this.enhancedNotes = const [],
  });

  final ArmorStatName stat;
  final List<String> baseEffects;
  final List<ScalingBenefit> baseScaling;
  final List<ScalingBenefit> enhancedScaling;
  final List<String> enhancedNotes;
}

final Map<ArmorStatName, StatBenefitDefinition> statBenefits = {
  ArmorStatName.health: const StatBenefitDefinition(
    stat: ArmorStatName.health,
    baseScaling: [
      ScalingBenefit(
        template: '+{v} HP healing when picking up an Orb of Power',
        max: 70,
      ),
      ScalingBenefit(template: '+{v}% flinch resistance', max: 10),
    ],
    enhancedScaling: [
      ScalingBenefit(template: '+{v} shield capacity vs combatants', max: 20),
      ScalingBenefit(template: 'shields start recharging {v}% sooner', max: 25),
      ScalingBenefit(
        template: '{v}% less time to fully recharge shields',
        max: 50,
      ),
    ],
  ),
  ArmorStatName.melee: const StatBenefitDefinition(
    stat: ArmorStatName.melee,
    baseEffects: [
      'Improves melee ability base cooldown',
      'Improves melee energy gained from external sources and regen scalars',
    ],
    enhancedScaling: [
      ScalingBenefit(template: '+{v}% melee ability damage', max: 30),
    ],
    enhancedNotes: ['Applies to powered, unpowered, and glaive melees'],
  ),
  ArmorStatName.grenade: const StatBenefitDefinition(
    stat: ArmorStatName.grenade,
    baseEffects: [
      'Improves grenade ability base cooldown',
      'Improves grenade energy gained from external sources and regen scalars',
    ],
    enhancedScaling: [
      ScalingBenefit(template: '+{v}% grenade ability damage', max: 65),
    ],
  ),
  ArmorStatName.superStat: const StatBenefitDefinition(
    stat: ArmorStatName.superStat,
    baseEffects: [
      'Improves Super energy gained from damaging targets (base cooldown unchanged)',
      'Improves Super energy gained from external sources and regen scalars',
    ],
    enhancedScaling: [
      ScalingBenefit(template: '+{v}% Super ability damage', max: 45),
    ],
  ),
  ArmorStatName.classStat: const StatBenefitDefinition(
    stat: ArmorStatName.classStat,
    baseEffects: [
      'Improves class ability base cooldown',
      'Improves class energy gained from external sources and regen scalars',
    ],
    enhancedScaling: [
      ScalingBenefit(
        template: '{v} HP overshield on casting your class ability (PvE)',
        max: 40,
      ),
      ScalingBenefit(template: '{v} HP overshield on cast in PvP', max: 10),
    ],
    enhancedNotes: [
      "Overshield duration is tied to the ability's cooldown length",
    ],
  ),
  ArmorStatName.weapons: const StatBenefitDefinition(
    stat: ArmorStatName.weapons,
    baseEffects: ['Improves weapon reload and handling speeds'],
    baseScaling: [
      ScalingBenefit(
        template:
            '+{v}% damage vs minor and major combatants (primary and special)',
        max: 15,
      ),
      ScalingBenefit(
        template: '+{v}% damage vs minor and major combatants (heavy)',
        max: 10,
      ),
    ],
    enhancedScaling: [
      ScalingBenefit(
        template: '+{v}% damage vs bosses (primary and special, PvE)',
        max: 15,
      ),
      ScalingBenefit(
        template: '+{v}% damage vs bosses (heavy, PvE)',
        max: 10,
      ),
      ScalingBenefit(
        template: '+{v}% weapon damage vs Guardians (PvP)',
        max: 6,
      ),
    ],
    enhancedNotes: [
      'Ammo bricks have a chance to contain more ammo than normal',
    ],
  ),
};

/// 9.7.0: damaging class abilities that additionally scale with Class above 100.
const List<String> classStatScalingAbilities = [
  'Shieldburst',
  'Ascension',
  "Threaded Specter",
  "Drengr's Lash",
];

const List<String> abilityEconomyNotes = [
  'Max passive ability-cooldown bonus from stats reduced by ~20% (9.7.0)',
  'Max active ability-cooldown bonus reduced from 190% to 125% (9.7.0)',
  'Super energy from boss damage reduced by 60% (9.7.0)',
  'Grenade/Melee stat boss-damage bonuses are additive, not multiplicative (9.7.0)',
];

double _interpolate(num value, num rangeStart, num max) {
  final clamped = (value - rangeStart).clamp(0, 100);
  return (clamped / 100) * max;
}

String _renderBenefit(ScalingBenefit benefit, num amount) {
  final rendered = amount.toStringAsFixed(benefit.precision);
  return benefit.template.replaceAll('{v}', rendered);
}

/// Human-readable benefit lines for a stat at [value] (0–200 scale).
/// Enhanced benefits are omitted at or below [enhancedThreshold].
List<String> computeBenefitsAt(ArmorStatName stat, num value) {
  final definition = statBenefits[stat]!;
  final lines = <String>[...definition.baseEffects];
  for (final benefit in definition.baseScaling) {
    lines.add(
      _renderBenefit(benefit, _interpolate(value, 0, benefit.max)),
    );
  }
  if (value <= enhancedThreshold) return lines;
  for (final benefit in definition.enhancedScaling) {
    lines.add(
      _renderBenefit(benefit, _interpolate(value, 100, benefit.max)),
    );
  }
  lines.addAll(definition.enhancedNotes);
  return lines;
}
